### DISCUSSION TITLE
does ripgrep's regex engine follow a specification? and is it compatible with .NET's regex engine?

### DISCUSSION THREAD

**Discussion by Akash-Mair** [BurntSushi/ripgrep#2602] [Q&A] [answered]
Hey regex noob here,

Im using both ripgrep and the `Regex` class from dotnet. I want to ensure that both are matching on the same input. Do you know if the `regex` crate that ripgrep uses is compatible with the regex engine `dotnet` uses?

Thank you :D

**Comment by BurntSushi** [comment, marked as answer]
It is not. AFAIK, .NET does not follow any spec. And ripgrep uses [Rust's `regex` crate](https://github.com/rust-lang/regex), and that doesn't follow any spec either. In general, there are only two broadly applicable regex specifications: [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html) and [ECMAScript](https://tc39.es/ecma262/#sec-regexp-regular-expression-objects). AFAIK, .NET allows you to opt into ECMAScript if you wanted to. But both specifications require implementing features that are not known how to implement efficiently (such as look-around). In contrast, Rust's regex crate follows the [RE2](https://github.com/google/re2) tradition and specific eschews those features so that it can use finite automata as an implementation technique and guarantee worst case linear time searches.

Rust's regex crate syntax is documented here: https://docs.rs/regex/latest/regex/#syntax

You can also browse through the `rebar` benchmark suite for regex engines (which includes Rust's regex crate and .NET). Its ["test" benchmarks](https://github.com/BurntSushi/rebar/tree/49527287944f5b8d80f9a5631e78af8e3eaef085/benchmarks/definitions/test/unicode) specifically expose some differences. Here is [one example of a difference](https://github.com/BurntSushi/rebar/blob/49527287944f5b8d80f9a5631e78af8e3eaef085/benchmarks/definitions/test/unicode/word.toml#L90-L91).

Overall, if you stick to "simple" regexes, then they should generally behave the same. But there are likely a number of corner cases where behavior differs. And of course, .NET supports more stuff that Rust's regex crate. If you need to deal with that, then you can opt into using PCRE2 in ripgrep with the `-P/--pcre2` flag. But there's no guarantee that PCRE2 matches what .NET does either.

**Reply by Akash-Mair** [reply]
Thank you so much!

**Reply by ltrzesniewski** [reply]
Just in case you need it, you can also use PCRE2 from .NET - I made [PCRE.NET](https://github.com/ltrzesniewski/pcre-net) for this purpose.

**Reply by BurntSushi** [reply]
It would also be possible to use Rust's regex crate from .NET. It exposes a C API: https://github.com/rust-lang/regex/tree/master/regex-capi/ctest

I don't think anyone has written an ffi binding for .NET for it though. It probably wouldn't be too hard.
